### PR TITLE
Fix for issue #3158

### PR DESCRIPTION
**Pull request by unp1**
Fix for issue #3158
- The program variable replacer also needs to consider lefthandside's of elementary updates
- Avoid collisions with names occurring in the namespace (and not only inside programs)
- Introduce new program variables properly in loop invariant rules
- Minor clean ups

### CODE REVIEW

**Comment by unp1** [comment]
This PR will break saved proofs as the loop scope taclets needed to be modified

File: key.core/src/main/resources/de/uka/ilkd/key/proof/rules/loopScopeRules.key [hunk 1]
```diff
@@ -62,6 +62,13 @@
 
         "Invariant Preserved and Used":
             \replacewith(
+              \modality{#dia}{
+                    #typeof(#heapBefore_LOOP) #heapBefore_LOOP;
+                    #typeof(#savedHeapBefore_LOOP) #savedHeapBefore_LOOP;
+                    #typeof(#permissionsBefore_LOOP) #permissionsBefore_LOOP;
+                    #typeof(#permissionsBefore_LOOP) #permissionsBefore_LOOP;
```
Comment (FliegendeWurst) [inline key.core/src/main/resources/de/uka/ilkd/key/proof/rules/loopScopeRules.key:69]: Why is this heap duplicated?

File: key.core/src/main/resources/de/uka/ilkd/key/proof/rules/loopScopeRules.key [hunk 2]
```diff
@@ -125,6 +133,13 @@
 
         "Invariant Preserved and Used":
             \replacewith(
+              \modality{#box}{
+                                #typeof(#heapBefore_LOOP) #heapBefore_LOOP;
+                                #typeof(#savedHeapBefore_LOOP) #savedHeapBefore_LOOP;
+                                #typeof(#permissionsBefore_LOOP) #permissionsBefore_LOOP;
+                                #typeof(#permissionsBefore_LOOP) #permissionsBefore_LOOP;
```
Comment (FliegendeWurst) [inline key.core/src/main/resources/de/uka/ilkd/key/proof/rules/loopScopeRules.key:140]: Same here

File: key.core/src/main/resources/de/uka/ilkd/key/proof/rules/loopScopeRules.key (hunk 1 above)
Comment (unp1) [reply key.core/src/main/resources/de/uka/ilkd/key/proof/rules/loopScopeRules.key]: Thanks! Is resolved in new commit.

**Comment by unp1** [comment]
The reduced test coverage is an unrelated file occurs because the exception is no longer thrown and hence the corresponding catch block in DiskFileRepo is no longer reached. That is expected behavior and should be ignored.

**Comment by unp1** [comment]
Will resume checks after next push

**Comment by mattulbrich** [comment]
The comment I made in #3158 seems to be fixed. "a" is now "a_1":
image

**Review by mattulbrich — APPROVED** [review]
This fixes the problem reported in the issue.

We haven't had these modalities introducing new program variables in a while, ... but if it works ... Would a direct addition to the namespace also be possible?

It's a bit strange that h can be referred to from the poststate of <>, but not toplevel. It is also actually not allowed in KeY. I tried a cut with
```
\<{ int x = 42; }\>x=42
```
which is a syntax error.

**Comment by unp1** [comment]
> This fixes the problem reported in the issue.
> 
> We haven't had these modalities introducing new program variables in a while, ... but if it works ... Would a direct addition to the namespace also be possible?

Sadly that is not as easy and it would require non-trivial changes to the taclet application process. The time when the variables are created in the taclet app is not the same time when the taclet app is applied. Hence, we need to book keep which variables might need to be replaced when applying such a taclet etc. 

But I agree it would be nice not to have to split the introduction of new variables in two steps.

> 
> It's a bit strange that h can be referred to from the poststate of <>, but not toplevel. It is also actually not allowed in KeY. I tried a cut with
> 
> ```
> \<{ int x = 42; }\>x=42
> ```
> 
> which is a syntax error.

I remember some discussions (years ago) that we wanted to allow that again (it was allowed pre KeY 2.0), as a simple way to introduce new program variables.

**Comment by WolframPfeifer** [comment]
I would favor this PR for the 2.12.0 release over #3243, since I think it is a more complete solution to the naming problems. Are there other disadvantages besides the broken reloading of older proofs? Because we broke that anyway due to the changed integer defaults and the changed OSS ruleset ...

What do you think?

@unp1
@mattulbrich

**Comment by mattulbrich** [comment]
The feature of introducing new variables like `x` in `<int x = 42>x >0` is not supported outside this use case. Is it clear that we want to have that feature at all? I would argue that the scope of x is local to the modality. A new variable should be added otherwise. ...

I seem to remember that that was a reason to split this PR

**Comment by unp1** [comment]
> The feature of introducing new variables like `x` in `<int x = 42>x >0` is not supported outside this use case. Is it clear that we want to have that feature at all? I would argue that the scope of x is local to the modality. A new variable should be added otherwise. ...
> 
> I seem to remember that that was a reason to split this PR

Hi,
 it might make sense to use this fix for the moment as w/o this fix two things are problematic:

* parsing: as different variables of different types cannot be disambiguated
* sequent comprehension: different variables with same names used

with this patch there is still the parsing problem due to the scoping, but someone just trying to understand the sequent will not be confused.

Best regards,
Richard

**Comment by WolframPfeifer** [comment]
> > The feature of introducing new variables like `x` in `<int x = 42>x >0` is not supported outside this use case. Is it clear that we want to have that feature at all? I would argue that the scope of x is local to the modality. A new variable should be added otherwise. ...
> > I seem to remember that that was a reason to split this PR
> 
> Hi, it might make sense to use this fix for the moment as w/o this fix two things are problematic:
> 
>     * parsing: as different variables of different types cannot be disambiguated
> 
>     * sequent comprehension: different variables with same names used
> 
> 
> with this patch there is still the parsing problem due to the scoping, but someone just trying to understand the sequent will not be confused.
> 
> Best regards, Richard

I am a bit confused now: By "this fix", you mean this PR (#3224) or the cherry-picked one (#3243)?

Since that seems to lead to a longer discussion, we could also consider keeping that status quo. I do not think it is that critical, since we have had that bug for years now ...

**Comment by unp1** [comment]
Hi, 

> I am a bit confused now: By "this fix", you mean this PR (#3224) or the cherry-picked one (#3243)?

By this fix I meant this PR.

I am fine with either decision.

Best regards, 
Richard

**Pull request merged**